### PR TITLE
"detector recall" command is amended with precision and class_id stats

### PR DESCRIPTION
**Pull request by IlyaOvodov**
"detector recall" command is amended with precision and class_id staistics:

1. "detector recall" command now takes parameter --thresh (default = 0) and include statistics for detections with class prob > thresh (like in "detector test" command)
2. added statistics on detections (precision and IOU)
3. added statistics taking into account detected class correctness (both for labels and detections): IOU, recall, precision with correct class_id
4. output format is changed to make all this information readable

### CODE REVIEW

**Comment by IlyaOvodov** [comment]
Oops! This PL does not compile without previous one. Waiting for #741 merge

**Comment by AlexeyAB** [comment]
I had once made minor changes to `detector recall` and this caused a lot of questions: why does new commit show better/worse accuracy than old commit, what is broken? why does your repo show better/worse accuracy than original repo?  ...

For this reason, I prefer to add new functions instead of change old functions.
So there is already new `detector map` that calculates standard widely used indicators AP and **mAP (AP50)** that are used in general competitions: Pascal VOC, MS COCO, ImageNet.

Also, yes, draw_detection should printf or all Classes id with prob>thresh or Best class id, but for accuracy indicators the use of `best_class` isn't always the best solution, if we are using multi-label classification, when 1 anchors can detect many overlaped objects with different classes for datasets like Open Images Dataset: https://pjreddie.com/media/files/papers/YOLOv3.pdf

> This formulation helps when we move to more complex
> domains like the Open Images Dataset [7]. In this dataset
> there are many overlapping labels (i.e. Woman and Person).
> Using a softmax imposes the assumption that each box has
> exactly one class which is often not the case. **A multilabel
> approach better models the data**.

Also `struct detection;` is used for Python C-wrappers, so changing of this `sturct` will broke darknet.py examples. So I try to keep unchanged basic structures.

**Comment by AlexeyAB** [comment]
> Do you think it is worth reworking it to fix (a), i.e. problems with change of basic structure, to merge it?

Yes, you can make `int best_class = calloc(num, sizeof(int));` as local array in the function `draw_detections_v3()`.

Also some people make own parser that parses output file `result.txt` that is got by using such command:
`darknet.exe detector test data/voc.data yolo-voc.cfg yolo-voc.weights -dont_show < data/train.txt > result.txt`
So ordering of output willn't broke such parsers. But adding `x,y,w,h`-coordinates can broke it. So better to add some flag that enables output absolute coordinates of detected objects. For example:
* without flag: no coordinate
* `-out_coords` - output absolute coordinates

I don't know should we output coords only for the best_class or for all classes of multilabel detection.

----

> PS. Another question, not to write twice. Do you know some way to get a learning curve automatically while learning (see picture https://hsto.org/files/5dc/7ae/7fa/5dc7ae7fad9d4e3eb3a484c58bfc1ff5.png from your Readme)

Curve of Training loss is drawed by default, just don't use flag `-dont_show`.
Probably I'll add curve of Validation dataset, but for good large training dataset - the overfitting usually occurs on very high number of iterations due to data augmentation. And there is more important to check Early Stopping Point by mAP on Validation dataset than check by loss on Validation dataset.
Also probably there is some reason to get mAP, for example, for each 1000 or 10 000 iterations and draw it on the loss chart.

chart_singlescale_832

**Comment by IlyaOvodov** [comment]
> Curve of Training loss is drawed by default, just don't use flag -dont_show.
Of cause I've seen this graph and i meant validation curve for early stop (like in a picture I've mentioned)

**Comment by IlyaOvodov** [comment]
This PR was also changed to preserve old output . New vershoin is switching on with -ext_output cmd line key.

**Comment by AlexeyAB** [comment]
@cenit I think this feature isn't required now, since all widly used indicators are in the `detector map`

**Comment by cenit** [comment]
shall you just close the PR in this case? Or do you prefer to keep it open for future work?
